src/rust_connection/inner.rs

```diff
@@ -323,18 +207,17 @@ where
     /// reply to be received. However, this approach does not work for requests without errors:
     /// Success is indicated by the absence of an error.
     ///
-    /// Thus, this function ensures that a reply with a higher sequence number will be received.
-    /// Since the X11 server handles requests in-order, if the reply to a later request is
-    /// received, this means that the earlier request did not fail.
-    pub(crate) fn prepare_check_for_reply_or_error(
-        &mut self,
-        sequence: SequenceNumber,
-    ) -> Result<(), std::io::Error> {
+    /// Thus, this function returns true if a sync is necessary to ensure that a reply with a
+    /// higher sequence number will be received. Since the X11 server handles requests in-order,
+    /// if the reply to a later request is received, this means that the earlier request did not
+    /// fail.
+    pub(crate) fn prepare_check_for_reply_or_error(&mut self, sequence: SequenceNumber) -> bool {
         if self.next_reply_expected < sequence {
-            self.send_sync()?;
+            true
+        } else {
+            assert!(self.next_reply_expected >= sequence);
```

Comment:
I think this `assert!` can be removed, it has just been checked by the `if`

Reply:
Yup, right. And then the whole `if` becomes redundant.

Updated.